File: 프로그래머스/3/132266. 부대복귀/부대복귀.py
from collections import deque

def solution(n, roads, sources, destination):
    # 1) 인접 리스트
    graph = [[] for _ in range(n + 1)]
    for a, b in roads:
        graph[a].append(b)
        graph[b].append(a)

    # 2) destination 기준 BFS
    dist = [-1] * (n + 1)
    dist[destination] = 0
    q = deque([destination])

    while q:
        cur = q.popleft()
        nd = dist[cur] + 1
        for nxt in graph[cur]:
            if dist[nxt] == -1:
                dist[nxt] = nd
                q.append(nxt)

    # 3) sources 순서대로 결과
    return [dist[s] for s in sources]

File: 프로그래머스/3/132266. 부대복귀/부대복귀.kt
import java.util.ArrayDeque

class Solution {
    fun solution(n: Int, roads: Array<IntArray>, sources: IntArray, destination: Int): IntArray {
        // 1) 인접 리스트
        val graph = Array(n + 1) { IntArray(0) }

        // roads가 최대 50만이라 ArrayList<Int>로도 되지만, 성능 위해 2-pass로 IntArray 구성
        val deg = IntArray(n + 1)
        for (e in roads) {
            val a = e[0]
            val b = e[1]
            deg[a]++
            deg[b]++
        }

        val temp = Array(n + 1) { IntArray(deg[it]) }
        val idx = IntArray(n + 1)

        for (e in roads) {
            val a = e[0]
            val b = e[1]
            temp[a][idx[a]++] = b
            temp[b][idx[b]++] = a
        }

        for (i in 1..n) {
            graph[i] = temp[i]
        }

        // 2) destination 기준 BFS 거리
        val dist = IntArray(n + 1) { -1 }
        dist[destination] = 0

        val q = ArrayDeque<Int>()
        q.add(destination)

        while (q.isNotEmpty()) {
            val cur = q.removeFirst()
            val nextDist = dist[cur] + 1

            for (nxt in graph[cur]) {
                if (dist[nxt] == -1) {
                    dist[nxt] = nextDist
                    q.add(nxt)
                }
            }
        }

        // 3) sources 순서대로 답 반환
        val answer = IntArray(sources.size)
        for (i in sources.indices) {
            answer[i] = dist[sources[i]]
        }
        return answer
    }
}